stats/registry.go

```diff
@@ -0,0 +1,50 @@
+package stats
+
+import (
+	"fmt"
+	"sync"
+)
+
+var errFmtMetricExists = "fatal: metric %q already exists"
+
+// Registry tracks metrics and reporters
+type Registry struct {
+	sync.Mutex
+	// here we use just the metric name as key. it does not include any prefix
+	// this means that technically we're more strict here then needed around naming conflicts:
+	// it's possible to have metrics with the same name, but a different prefix.
+	// we still complain about that
+	metrics map[string]GraphiteMetric
+}
+
+func NewRegistry() *Registry {
+	return &Registry{
+		metrics: make(map[string]GraphiteMetric),
+	}
+}
+
+func (r *Registry) add(name string, metric GraphiteMetric) GraphiteMetric {
+	r.Lock()
+	if _, ok := r.metrics[name]; ok {
+		panic(fmt.Sprintf(errFmtMetricExists, name))
```

Comment:
I think AddOrGet is a good idea.
However, the pre-existing metric may be a different type.  E.g. trying to register a Bool, but there's already a Meter32 or something. There's not much useful we can do in this scenario (we could keep running but without the instrumentation, which I think is pretty dangerous and we'd have to make the code more complicated to support a graphite metric type which is a no-op). So if the pre-existing one is different type, i think a panic is still warranted.

if pre-existing is the same type, then we can totally just return that.